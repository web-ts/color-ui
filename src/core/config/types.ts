import { Component } from "solid-js";
import { I18nProvider } from "../i18n/types";

export type ColorOptions = {
  primary: string;
  secondary: string;
  tertiary: string;
  success: string;
  info: string;
  warning: string;
  danger: string;
  background: string;
  text: {
    light: string;
    dark: string;
  };
};

export type ColorUIOptions = {
  colors: {
    light: ColorOptions;
    dark: ColorOptions;
  };
  icons: {
    spinner: string;
    close: string;
    menu: string;
    info: string;
    success: string;
    warning: string;
    danger: string;
  };
  i18nProvider?: I18nProvider;
  iconProvider: {
    component: Component<any>;
    keyPropName: string;
    defaultProps: Record<string, any>;
  };
};
