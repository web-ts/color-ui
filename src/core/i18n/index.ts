import { I18nProvider } from "./types";
import defaultProvider from "./providers/default-provider";

let provider: I18nProvider = defaultProvider();

export function setProvider(newProvider: I18nProvider) {
  provider = newProvider;
}

export default function () {
  return provider;
}
