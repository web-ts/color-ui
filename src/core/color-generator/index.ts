import { ColorOptions } from "../config/types";
import { Color, ColorSet, ColorTextSet } from "./types";

/**
 * Convert hex color to RGB
 */
export function hexToRgb(hex: string): Color {
  if (typeof hex !== "string") {
    throw new Error("hex must be a string");
  }

  const regex = /^#([0-9a-f]{3}|[0-9a-f]{6})$/i;
  if (!regex.test(hex)) {
    throw new Error("hex must be a valid hex color");
  }

  let r = hex.slice(1, 3);
  let g = hex.slice(3, 5);
  let b = hex.slice(5, 7);

  const intR = parseInt(r, 16);
  const intG = parseInt(g, 16);
  const intB = parseInt(b, 16);

  return { r: intR, g: intG, b: intB };
}

export function lightenColor(color: Color, amount: number): Color {
  const { r, g, b } = color;

  if (amount < 0 || amount > 1) {
    throw new Error("amount must be between 0 and 1");
  }

  const lightenAmount = Math.floor(amount * 255);

  return {
    r: r + lightenAmount > 255 ? 255 : r + lightenAmount,
    g: g + lightenAmount > 255 ? 255 : g + lightenAmount,
    b: b + lightenAmount > 255 ? 255 : b + lightenAmount,
  };
}

export function darkenColor(color: Color, amount: number): Color {
  const { r, g, b } = color;

  if (amount < 0 || amount > 1) {
    throw new Error("amount must be between 0 and 1");
  }

  const darkenAmount = Math.floor(amount * 255);

  return {
    r: r - darkenAmount < 0 ? 0 : r - darkenAmount,
    g: g - darkenAmount < 0 ? 0 : g - darkenAmount,
    b: b - darkenAmount < 0 ? 0 : b - darkenAmount,
  };
}

export function colorSet(color: Color): ColorSet {
  return {
    base: color,
    light: lightenColor(color, 0.1),
    lighter: lightenColor(color, 0.2),
    lightest: lightenColor(color, 0.3),
    dark: darkenColor(color, 0.1),
    darker: darkenColor(color, 0.2),
    darkest: darkenColor(color, 0.3),
  };
}

export function accesibleTextColor(
  color: Color,
  text: { light: Color; dark: Color }
): Color {
  let r = color.r / 255;
  let g = color.g / 255;
  let b = color.b / 255;
  r = r <= 0.03928 ? r / 12.92 : Math.pow((r + 0.055) / 1.055, 2.4);
  g = g <= 0.03928 ? g / 12.92 : Math.pow((g + 0.055) / 1.055, 2.4);
  b = b <= 0.03928 ? b / 12.92 : Math.pow((b + 0.055) / 1.055, 2.4);
  let luminance = 0.2126 * r + 0.7152 * g + 0.0722 * b;

  if (luminance > 0.21) {
    return text.dark;
  } else {
    return text.light;
  }
}

export function colorTextSet(
  colorSet: ColorSet,
  text: { light: Color; dark: Color }
): ColorTextSet {
  const { base, light, lighter, lightest, dark, darker, darkest } = colorSet;

  return {
    base: { color: base, text: accesibleTextColor(base, text) },
    light: { color: light, text: accesibleTextColor(light, text) },
    lighter: { color: lighter, text: accesibleTextColor(lighter, text) },
    lightest: { color: lightest, text: accesibleTextColor(lightest, text) },
    dark: { color: dark, text: accesibleTextColor(dark, text) },
    darker: { color: darker, text: accesibleTextColor(darker, text) },
    darkest: { color: darkest, text: accesibleTextColor(darkest, text) },
  };
}

export function colorTextSetToStyleObject(
  colorTextSet: ColorTextSet,
  name: string
): Record<string, string> {
  const styleObject: Record<string, string> = {};

  Object.entries(colorTextSet).forEach(([key, value]) => {
    const { color, text } = value;
    const { r, g, b } = color;
    const { r: tR, g: tG, b: tB } = text;

    styleObject[`--color-${name}-${key}`] = `${r}, ${g}, ${b}`;
    styleObject[`--color-${name}-${key}-text`] = `${tR}, ${tG}, ${tB}`;
  });

  return styleObject;
}

export function colorOptionsToStyleObject(
  colorOptions: ColorOptions
): Record<string, string> {
  const styleObject: Record<string, string> = {};

  const textColors = {
    light: hexToRgb(colorOptions.text.light),
    dark: hexToRgb(colorOptions.text.dark),
  };

  Object.entries(colorOptions).forEach(([key, value]) => {
    if (typeof value === "string") {
      const color = hexToRgb(value);
      const colorSetResult = colorSet(color);
      const colorTextSetResult = colorTextSet(colorSetResult, textColors);
      const colorTextSetStyleObject = colorTextSetToStyleObject(
        colorTextSetResult,
        key
      );

      Object.assign(styleObject, colorTextSetStyleObject);
    }
  });

  return styleObject;
}

export function colorOptionsToCssString(colorOptions: ColorOptions) {
  return `${Object.entries(colorOptionsToStyleObject(colorOptions)).reduce(
    (acc, [key, value]) => `${acc}  ${key}: ${value};\n`,
    "* {\n"
  )} }`;
}
