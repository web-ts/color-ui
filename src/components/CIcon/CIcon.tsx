import { getConfig } from "@/core/config";
import { Component } from "solid-js";
import { Dynamic } from "solid-js/web";

export const CIcon: Component<{ children: string } & Record<string, any>> = (
  props
) => {
  const iconProvider = getConfig().iconProvider;

  return (
    <Dynamic
      component={iconProvider.component}
      {...{
        ...iconProvider.defaultProps,
        ...props,
        [iconProvider.keyPropName]: props.children,
      }}
    />
  );
};
