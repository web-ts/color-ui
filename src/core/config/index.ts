import { DEFAULT_COLOR_UI_OPTIONS } from '../constants';
import { ColorUIOptions } from './types';

let config: ColorUIOptions = { ...DEFAULT_COLOR_UI_OPTIONS };

export function getConfig() {
  return config;
}

export function createColorUI(newConfig: Partial<ColorUIOptions>) {
  config = { ...config, ...newConfig };
}
