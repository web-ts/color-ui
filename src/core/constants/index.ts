import { ColorUIOptions } from "../config/types";

export const DEFAULT_COLORS = {
  light: {
    primary: "#3498db", // a light blue color often used for links or accents
    secondary: "#7f8c8d", // a medium gray color for secondary text or borders
    tertiary: "#2ecc71", // a light green color for additional accents or success messages
    success: "#27ae60", // a darker green color for indicating success or confirmation
    info: "#2980b9", // a darker blue color for information or data
    warning: "#f39c12", // a yellow-orange color for warning or cautionary messages
    danger: "#e74c3c", // a red color for indicating danger or errors
    background: "#ecf0f1", // a light gray color for the background
    text: {
      light: "#fefefe", // a white color for text on a dark background
      dark: "#090909", // a black color for text on a light background
    },
  },
  dark: {
    primary: "#1abc9c", // a turquoise color for links or accents in a dark theme
    secondary: "#bdc3c7", // a lighter gray color for secondary text or borders
    tertiary: "#9b59b6", // a purple color for additional accents or success messages
    success: "#2ecc71", // a lighter green color for indicating success or confirmation
    info: "#3498db", // a brighter blue color for information or data
    warning: "#f1c40f", // a brighter yellow-orange color for warning or cautionary messages
    danger: "#e74c3c", // the same red color as in the light theme for indicating danger or errors
    background: "#2c3e50", // a dark blue color for the background
    text: {
      light: "#fefefe", // the same white color as in the light theme for text on a dark background
      dark: "#090909", // the same black color as in the light theme for text on a light background
    },
  },
};

export const DEFAULT_ICONS = {
  spinner: "line-md:loading-twotone-loop",
  close: "ic:round-close",
  menu: "ic:round-menu",
  info: "ic:round-info",
  success: "ic:round-check-circle",
  warning: "ic:round-warning",
  danger: "ic:round-dangerous",
};

export const DEFAULT_COLOR_UI_OPTIONS: ColorUIOptions = {
  colors: DEFAULT_COLORS,
  icons: DEFAULT_ICONS,
  iconProvider: {
    component: null as any,
    keyPropName: "icon",
    defaultProps: {},
  },
};
