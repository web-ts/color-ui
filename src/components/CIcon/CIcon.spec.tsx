import { it, expect, describe } from "vitest";

import { CIcon } from "./CIcon";
import { render } from "@solidjs/testing-library";
import { createColorUI } from "@/core/config";

describe("CIcon", () => {
  it("should render empty div if no provider is set", () => {
    const { container } = render(() => <CIcon>mdi:home</CIcon>);

    expect(container).toMatchSnapshot();
  });

  it("should render an icon for the specified provider", () => {
    createColorUI({
      iconProvider: {
        component: (props: { icon: string }) => <span>{props.icon}</span>,
        keyPropName: "icon",
        defaultProps: {},
      },
    });

    const { container } = render(() => <CIcon>mdi:home</CIcon>);

    expect(container).toMatchSnapshot();
  });
});
