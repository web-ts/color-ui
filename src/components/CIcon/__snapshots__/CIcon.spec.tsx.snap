// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CIcon > should render 1`] = `<div />`;

exports[`CIcon > should render an icon for the specified provider 1`] = `
<div>
  <span>
    mdi:home
  </span>
</div>
`;

exports[`CIcon > should render empty div if no provider is set 1`] = `<div />`;
